include/nil/actor/container/merkle/tree.hpp

```diff
@@ -116,6 +115,82 @@ namespace nil {
                     }
                     return *ret_p;
                 }
+
+                template<
+                    typename T, std::size_t Arity, typename LeafIterator,
+                    std::enable_if_t<crypto3::hashes::is_poseidon<typename T::hash_type>::value, bool> = true
+                >
+                nil::crypto3::containers::detail::merkle_tree_impl<T, Arity> make_merkle_tree(LeafIterator first,
+                                                                                              LeafIterator last) {
+//                    We can compare it with one-thread version with multi-thread
+//                    nil::crypto3::containers::detail::make_merkle_tree<T, Arity, LeafIterator>(first, last);
+
+                    typedef T node_type;
+                    typedef typename node_type::hash_type hash_type;
+                    typedef typename node_type::value_type value_type;
+
+                    std::size_t number_leaves = std::distance(first, last);
+                    nil::crypto3::containers::detail::merkle_tree_impl<T, Arity> ret(number_leaves);
+                    std::size_t core_count = nil::actor::smp::count;
+                    ret.resize(ret.complete_size());
+
+                    auto ret_p = make_foreign(&ret);
+                    std::size_t parallels = std::min(core_count-1, number_leaves);
+                    std::vector<nil::actor::future<>> fut;
+                    for (auto i = 0; i < parallels; ++i) {
+                        auto begin = i * (number_leaves / parallels);
+                        auto end = (i + 1) * (number_leaves / parallels);
+                        if (i == parallels - 1) {
+                            end = number_leaves;
+                        }
+                        fut.emplace_back(nil::actor::smp::submit_to(i+1, [begin, end, first, p = ret_p.get()]() {
+
+                            auto itr = first;
+                            for (auto i = begin; i < end; ++i) {
+                                auto leaf = *itr;
+                                (*p)[i] = (value_type)nil::crypto3::containers::detail::generate_poseidon_leaf_hash<hash_type>(leaf);
+                                itr++;
+                            }
+                            return nil::actor::make_ready_future<>();
+                        }));
+                        first += number_leaves / parallels;
+                    }
+                    nil::actor::when_all(fut.begin(), fut.end()).get();
+
+                    std::size_t row_size = ret.leaves() / Arity;
+                    typename nil::crypto3::containers::detail::merkle_tree_impl<T, Arity>::iterator it = ret.begin();
```

Comment:
Mb `auto` would work here?
```suggestion
                    auto it = ret.begin();
```

Reply:
Let's keep the full type here for now.